### PR TITLE
Redefine objective sense as a proper `IntEnum`

### LINKED ISSUE #993: Conflicting definitions of minimize
There are multiple definitions of "minimize" that are used in Pyomo. Crucially, we run into this issue:

```
>>> from pyomo.environ import minimize
>>> from pyomo.opt import ProblemSense
>>> ProblemSense.minimize == minimize
False
```

The `minimize` from `pyomo.environ` is being imported from `pyomo.core.kernel.objective`.

This can obviously cause some issues when checking objective sense.

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
There have been some longstanding inconsistencies regarding how we manage the definition of "objective sense" in Pyomo.  The results object uses an Enum with 3 values (minimize, maximize, and unknown).  In the modeling side, we make heavy use of two named constants (`minimize` and `maximize`).  These constants are just `int`s (1 and -1, respectively), and those constants are defined in `pyomo.core.kernel.objective` (causing one of the few inter-linkages between kernel and the AML).

This PR defines a new Enum (`pyomo.common.enums.ObjectiveSense`), and redefines the `minimize` and `maximize` constants to be the members of that Enum.  To preserve consistency / backwards compatibility, `ObjectiveSense` is an `IntEnum`, so the old constants / old usage does not need to change.  

`ProblemSense` is now an extension of `ObjectiveSense` that also admits `unknown`.  Unfortunately, due to the way that the old solvers are structured, it is not (easy) to remove the use of `unknown` when processing results from the old solver interfaces.  The compromise here is the development of `ExtendedEnumType`, which allows us to define extended (or derived) Enums that combine the enums from another class with additional locally-defined members - so the `minimize` and `maximize` members in `ProblemSense` are actually members of `ObjectiveSense`.

## Changes proposed in this PR:
- Create `ExtendedEnumType` metaclass to support extended/derived Enums
- Add `ObjectiveSense` enum (with `minimize` and `maximize` members)
- Redefine `ProblemSense` to be an extension of `ObjectiveSense`
- Remove references to `ProblemSense` from almost all of Pyomo
- Fix an error in processing the lower / upper bound in the SCIPAMPL interface

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/solvers/tests/mip/test_scip_solve_from_instance.baseline [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
     "Problem": [
         {
-            "Lower bound": -Infinity,
+            "Lower bound": 1.0,
```
Comment (mrmundt) [inline pyomo/solvers/tests/mip/test_scip_solve_from_instance.baseline:4]: How is this related? If not, no biggie, just curious.

File: pyomo/common/tests/test_enums.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        self.assertTrue(hasattr(ProblemSense, 'maximize'))
+
+    def test_call(self):
+        self.assertIs(ObjectiveSense(1), ObjectiveSense.minimize)
+        self.assertIs(ObjectiveSense(-1), ObjectiveSense.maximize)
+
+        self.assertIs(ObjectiveSense('minimize'), ObjectiveSense.minimize)
+        self.assertIs(ObjectiveSense('maximize'), ObjectiveSense.maximize)
+
+        with self.assertRaisesRegex(ValueError, "'foo' is not a valid ObjectiveSense"):
```
Comment (mrmundt) [inline pyomo/common/tests/test_enums.py:90]: It may be worth adding another check here for if someone tries to use something not in the set of [-1, 0, 1].

File: pyomo/solvers/tests/mip/test_scip_solve_from_instance.baseline (hunk 1 above)
Comment (jsiirola) [reply pyomo/solvers/tests/mip/test_scip_solve_from_instance.baseline:4]: It's not - apart from as I was removing use of `ProblemSense` I ran across this bug in the SCIP interface where we always reported the incumbent objective value as the upper bound (regardless of the optimization sense).  I fixed that (instead of preserving the bug and fixing it in a separate PR) - and then promptly realized that we never returned the opposing (dual) bound, and that we should.  So I fixed that, too.  That latter fix caused the change in the baseline.

**Review by mrmundt — APPROVED** [review]

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py [hunk 3]
```diff
@@ -374,9 +369,11 @@ def _postsolve(self):
             if len(results.solution) > 0:
                 results.solution(0).status = SolutionStatus.optimal
             try:
-                if results.problem.sense == ProblemSense.minimize:
+                if results.solver.primal_bound < results.solver.dual_bound:
                     results.problem.lower_bound = results.solver.primal_bound
+                    results.problem.upper_bound = results.solver.dual_bound
                 else:
+                    results.problem.lower_bound = results.solver.dual_bound
                     results.problem.upper_bound = results.solver.primal_bound
```
Comment (emma58) [inline pyomo/solvers/plugins/solvers/SCIPAMPL.py:377]: Can this go wrong if your bounds cross by a little bit?

**Review by emma58 — APPROVED** [review]
I like this! I have one question about the SCIP change below--just wondering if you could fall victim to numerical issues.

File: pyomo/solvers/plugins/solvers/SCIPAMPL.py (hunk 3 above)
Comment (jsiirola) [reply pyomo/solvers/plugins/solvers/SCIPAMPL.py:377]: Yes - you probably can.  This will be easily fixable in the new solver interfaces (where the results parser can have easier access to the original model and won't have to guess about the sense), but there isn't an obvious better solution here (unless the scip output log prints something about the sense - but I really don't want to put more dependence on the parsing of output logs).

File: pyomo/opt/results/problem.py [hunk 4]
```diff
@@ -12,19 +12,17 @@
 import enum
 from pyomo.opt.results.container import MapContainer
 
+from pyomo.common.deprecation import deprecated, deprecation_warning
```
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/opt/results/problem.py]: Nope.  Removed.

**Review by blnicho — APPROVED** [review]
